microcosm_pubsub/backoff.py

```diff
@@ -0,0 +1,66 @@
+"""
+Consumer backoff policy
+
+"""
+from abc import ABCMeta, abstractmethod
+from random import randint
+from six import add_metaclass
+
+
+# we cannot go over 12 hours
+MAX_BACKOFF_TIMEOUT = 60 * 60 * 12
+
+
+@add_metaclass(ABCMeta)
+class BackoffPolicy(object):
+
+    def __init__(self, visibility_timeout_seconds=None):
+        self.visibility_timeout_seconds = visibility_timeout_seconds
+
+    @abstractmethod
+    def compute_backoff_timeout(self, message, message_timeout):
+        pass
+
+    @classmethod
+    def choose_backoff_policy(cls, name):
+        for subclass in cls.__subclasses__():
+            if subclass.__name__ == name:
+                return subclass
+        raise Exception("No backoff policy configured with class name: {}".format(name))
+
+
+class NaiveBackoffPolicy(BackoffPolicy):
+    """
+    Naive backoff policy.
+
+    Uses a fixed timeout from the message or system default.
+
+    """
+    def compute_backoff_timeout(self, message, message_timeout):
+        backoff_timeout = message_timeout or self.visibility_timeout_seconds
+        # we can only set integer timeouts
+        return int(backoff_timeout) if backoff_timeout is not None else backoff_timeout
+
+
+class ExponentialBackoffPolicy(BackoffPolicy):
```

Comment:
This policy does (almost) textbook exponential backoff except:
 - It's always integral (SQS limitation)
 - It never returns `0` (seems pointless, but I could be wrong)
 - It avoids returns more than 12 hours (SQS limitation)